engine/lib/phases/phase_rewrite_control_flow.ml

```diff
@@ -0,0 +1,103 @@
+(* This phase rewrites: `if c {return a}; b` as `if c {return a; b} else {b}`
+   and does the equivalent transformation for pattern matchings. *)
+
+open! Prelude
+
+module Make (F : Features.T) =
+  Phase_utils.MakeMonomorphicPhase
+    (F)
+    (struct
+      let phase_id = Diagnostics.Phase.RewriteControlFlow
+
+      open Ast.Make (F)
+      module U = Ast_utils.Make (F)
+      module Visitors = Ast_visitors.Make (F)
+
+      module Error = Phase_utils.MakeError (struct
+        let ctx = Diagnostics.Context.Phase phase_id
+      end)
+
+      let has_return =
+        object (_self)
+          inherit [_] Visitors.reduce as super
+          method zero = false
+          method plus = ( || )
+
+          method! visit_expr' () e =
+            match e with Return _ -> true | _ -> super#visit_expr' () e
+        end
+
+      let rewrite_control_flow =
+        object (self)
+          inherit [_] Visitors.map as super
+
+          method! visit_expr () e =
+            match e.e with
+            | _ when not (has_return#visit_expr () e) -> e
+            (* Returns in loops will be handled by issue #196 *)
+            | Loop _ -> e
+            | Let _ -> (
+                (* Collect let bindings to get the sequence
+                   of "statements", find the first "statement" that is a
+                   control flow containing a return. Rewrite it.
+                *)
+                let stmts, final = U.collect_let_bindings e in
+                let inline_in_branch branch p stmts_after final =
+                  let branch_stmts, branch_final =
+                    U.collect_let_bindings branch
+                  in
+                  U.make_lets
+                    (branch_stmts @ ((p, branch_final) :: stmts_after))
+                    final
+                in
+                match
```

Comment:
I'm nitpicking a little, but maybe you should put that split_while in a let binding so that you can extract `stmts_before` before matching? (since you don't actually match on `stmts_before`)